streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -79,31 +82,32 @@ private void initInternal(final InternalProcessorContext context) {
     private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
         final Bytes binaryKey = cacheFunction.key(entry.key());
         final Windowed<Bytes> bytesKey = SessionKeySchema.from(binaryKey);
+        final byte[] newValueBytes = entry.newValue();
         if (flushListener != null) {
-            final byte[] newValueBytes = entry.newValue();
             final byte[] oldValueBytes = newValueBytes == null || sendOldValues ?
                 wrapped().fetchSession(bytesKey.key(), bytesKey.window().start(), bytesKey.window().end()) : null;
 
             // this is an optimization: if this key did not exist in underlying store and also not in the cache,
             // we can skip flushing to downstream as well as writing to underlying store
             if (newValueBytes != null || oldValueBytes != null) {
                 // we need to get the old values if needed, and then put to store, and then flush
-                wrapped().put(bytesKey, entry.newValue());
+                wrapped().put(bytesKey, newValueBytes);
 
                 final ProcessorRecordContext current = context.recordContext();
-                context.setRecordContext(entry.entry().context());
+                final ProcessorRecordContext cachedContext = entry.entry().context();
+                context.setRecordContext(cachedContext);
                 try {
                     flushListener.apply(
                         binaryKey.get(),
                         newValueBytes,
                         sendOldValues ? oldValueBytes : null,
-                        entry.entry().context().timestamp());
+                        dslUsage ? bytesKey.window().end() : cachedContext.timestamp());
```

Comment:
I see what you are saying. Atm, it does not make a difference, because the `FlushListener` calls:
```
context.forward(key, new Change<>(newValue, oldValue), To.all().withTimestamp(timestamp));
```
and this will set the context timestamp to the window-end timestamp before calling downstream processors.

As a matter of fact, I have the suspicion that we can actually remove the timestamp argument on the `FlushListener` -- it was added as part of KIP-258, but after some refactoring of our stores, I _think_ we can change it back (my idea was to hold back until KIP-258 is finished). For this case, the problem resolves naturally. Would it be ok if we clean this up as a follow up (to be sure we can remove the `timestamp` parameter again)?

Reply:
> Would it be ok if we clean this up as a follow up (to be sure we can remove the timestamp parameter again)?

Of course! Thanks for looking at it.